File: Transmitter/sender.py
#Pascal Schäfer
import socket
import hashlib
import os
import struct
import sys
import time


def execute_sendSleep(transmissionid,PORT,ipadress,buffer,filenameabs):


    # Definiere Konstanten
    BUFFER_SIZE = int(buffer)
    UDP_IP = ipadress     #'localhost'    127.0.0.1'
    UDP_PORT = int(PORT)   #5005

    # Wähle die zu sendende Datei
    filenamebase = os.path.basename(filenameabs)      #'example.txt'
    #/home/Pascal/Studium/Netze/Projekt1/Abgabe/NetzePS/Transmitter/testfiles/example1MB.txt
    #or just example.txt when file in same dir
    
    # Erstelle UDP-Socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Bestimme die maximale Sequenznummer

    max_seq = os.path.getsize(filenameabs) // BUFFER_SIZE
    if os.path.getsize(filenameabs) % BUFFER_SIZE != 0:
        max_seq += 1


    #Für letztes Paket mit md5
    max_seq +=1
    print(max_seq)
    # Sende das erste Paket mit der Dateiinformation
    trans_id = int(transmissionid) #1234  Wähle eine Transmission ID
    seq_num = 0
    filename_encoded = filenamebase.encode()
    if (len(filename_encoded)<8 | len(filename_encoded)>2048):
        print('\nuse filenameabs longer 8 and smaller 2048 bytes')
        sys.exit()


    header = struct.pack('!HLL', trans_id, seq_num, max_seq) + filename_encoded

    #start time transit
    starttransmit = time.time()
    print("Start sending...")
    sock.sendto(header, (UDP_IP, UDP_PORT))

    
    
    # Sende die Datenpakete
    with open(filenameabs, 'rb') as f:
        packets_sent=0
        for i in range(max_seq-1):
            data = f.read(BUFFER_SIZE)
            seq_num += 1
            packet = struct.pack('!HL', trans_id, seq_num) + data
            sock.sendto(packet, (UDP_IP, UDP_PORT))
            packets_sent+=1
            time.sleep(0.0000001*BUFFER_SIZE)

            #just for percentage display
            if(max_seq>=10):
                if (packets_sent % (max_seq//10) == 0):
                    percentage_sent = round(packets_sent/max_seq*100)
                    print(f'{percentage_sent}%')
        print(f'Transmission complete.')


    # Sende das letzte Paket mit dem MD5-Hash
    with open(filenameabs, 'rb') as f:
        data = f.read()

    md5 = hashlib.md5(data).digest()
    packet = struct.pack('!HL', trans_id, max_seq) + md5
    sock.sendto(packet, (UDP_IP, UDP_PORT))
    endtransmit = time.time()
    print(f'hash cal:{md5.hex()}')
    print(f'Transmission time:{endtransmit-starttransmit}')
    # Schließe den Socket
    
    sock.close()


def execute_sendACK(transmissionid,PORT,ipadress,buffer,filenameabs):


    # Definiere Konstanten
    BUFFER_SIZE = int(buffer)
    UDP_IP = ipadress     #'localhost'    127.0.0.1'
    UDP_PORT = int(PORT)   #5005

    # Wähle die zu sendende Datei
    filenamebase = os.path.basename(filenameabs)      #'example.txt'
    #/home/Pascal/Studium/Netze/Projekt1/Abgabe/NetzePS/Transmitter/testfiles/example1MB.txt
    #or just example.txt when file in same dir
    
    # Erstelle UDP-Socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Bestimme die maximale Sequenznummer

    max_seq = os.path.getsize(filenameabs) // BUFFER_SIZE
    if os.path.getsize(filenameabs) % BUFFER_SIZE != 0:
        max_seq += 1


    #Für letztes Paket mit md5
    max_seq +=1
    print(max_seq)
    # Sende das erste Paket mit der Dateiinformation
    trans_id = int(transmissionid) #1234  Wähle eine Transmission ID
    seq_num = 0
    filename_encoded = filenamebase.encode()
    if (len(filename_encoded)<8 | len(filename_encoded)>2048):
        print('\nuse filenameabs longer 8 and smaller 2048 bytes')
        sys.exit()


    header = struct.pack('!HLL', trans_id, seq_num, max_seq) + filename_encoded

    #start time transit
    starttransmit = time.time()
    print("Start sending...")
    sock.sendto(header, (UDP_IP, UDP_PORT))

    #wait for ack 

    received_ack,address = sock.recvfrom(6)
    ack_trans_id,ack_seq_num = struct.unpack('!HL',received_ack)
    if ack_trans_id != trans_id or ack_seq_num != seq_num:
        print('first packet transfer failed')
        sys.exit()



    
    # Sende die Datenpakete
    with open(filenameabs, 'rb') as f:
        packets_sent=0
        for i in range(max_seq-1):
            data = f.read(BUFFER_SIZE)
            seq_num += 1
            packet = struct.pack('!HL', trans_id, seq_num) + data
            sock.sendto(packet, (UDP_IP, UDP_PORT))
            packets_sent+=1
            received_ack,address = sock.recvfrom(6)
            ack_trans_id,ack_seq_num = struct.unpack('!HL',received_ack)
            if ack_trans_id != trans_id or ack_seq_num != seq_num:
                print(f"{seq_num} packet transfer failed")
                sys.exit()

            #just for percentage display
            if(max_seq>=10):
                if (packets_sent % (max_seq//10) == 0):
                    percentage_sent = round(packets_sent/max_seq*100)
                    print(f'{percentage_sent}%')
        print(f'Transmission complete.')


    # Sende das letzte Paket mit dem MD5-Hash
    with open(filenameabs, 'rb') as f:
        data = f.read()

    md5 = hashlib.md5(data).digest()
    packet = struct.pack('!HL', trans_id, max_seq) + md5
    sock.sendto(packet, (UDP_IP, UDP_PORT))
    endtransmit = time.time()
    print(f'hash cal:{md5.hex()}')
    print(f'Transmission time:{endtransmit-starttransmit}')
    # Schließe den Socket
    
    sock.close()


def execute_sendSlidingWindow(transmissionid,PORT,ipadress,buffer,filenameabs):


    # Definiere Konstanten
    WINDOW_SIZE= 10
    BUFFER_SIZE = int(buffer)
    UDP_IP = ipadress     #'localhost'    127.0.0.1'
    UDP_PORT = int(PORT)   #5005

    # Wähle die zu sendende Datei
    filenamebase = os.path.basename(filenameabs)      #'example.txt'
    #/home/Pascal/Studium/Netze/Projekt1/Abgabe/NetzePS/Transmitter/testfiles/example1MB.txt
    #or just example.txt when file in same dir
    
    # Erstelle UDP-Socket
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Bestimme die maximale Sequenznummer

    max_seq = os.path.getsize(filenameabs) // BUFFER_SIZE
    if os.path.getsize(filenameabs) % BUFFER_SIZE != 0:
        max_seq += 1


    #Für letztes Paket mit md5
    max_seq +=1
    print(max_seq)
    # Sende das erste Paket mit der Dateiinformation
    trans_id = int(transmissionid) #1234  Wähle eine Transmission ID
    seq_num = 0
    filename_encoded = filenamebase.encode()
    if (len(filename_encoded)<8 | len(filename_encoded)>2048):
        print('\nuse filenameabs longer 8 and smaller 2048 bytes')
        sys.exit()


    header = struct.pack('!HLL', trans_id, seq_num, max_seq) + filename_encoded

    #start time transit
    starttransmit = time.time()
    print("Start sending...")
    sock.sendto(header, (UDP_IP, UDP_PORT))

    #wait for ack 

    received_ack,address = sock.recvfrom(6)
    ack_trans_id,ack_seq_num = struct.unpack('!HL',received_ack)
    if ack_trans_id != trans_id or ack_seq_num != seq_num:
        print('first packet transfer failed')
        sys.exit()



    
    # Sende die Datenpakete
    packet_array = []
    with open(filenameabs, 'rb') as f:
        
        for i in range(max_seq-1):
            data = f.read(BUFFER_SIZE)
            seq_num += 1
            packet = struct.pack('!HL', trans_id, seq_num) + data
            packet_array.append(packet)
        
    

    # Sende die Datenpakete
    window_packets_counter = 0
    window_base = 0
    seq_num=1
    
    while seq_num < (max_seq-1):
        
        #sendet WINDOW_SIZE viele pakete
        while (seq_num < max_seq) & (window_packets_counter < WINDOW_SIZE):                
            
            
            #print(seq_num,max_seq,window_base)                 
            sock.sendto(packet_array[window_base], (UDP_IP, UDP_PORT))
            window_packets_counter+=1
            window_base+=1
            seq_num += 1
                

            
        received_ack,address = sock.recvfrom(6)
        ack_trans_id,ack_seq_num = struct.unpack('!HL',received_ack)
        #print(f"window base:{seq_num} seq_num={ack_seq_num}")
        if ack_trans_id != trans_id :
            print(f"{seq_num} packet transfer failed")
            #window_base=window_base-WINDOW_SIZE
            sys.exit()
                  
              
        if(ack_seq_num < seq_num):
            print(f"duplicate received: {ack_seq_num}")
            window_base=window_base - (window_base-ack_seq_num)
            seq_num= window_base+1
            
        
        window_packets_counter=0



    # Sende das letzte Paket mit dem MD5-Hash
    with open(filenameabs, 'rb') as f:
        data = f.read()

    md5 = hashlib.md5(data).digest()
    packet = struct.pack('!HL', trans_id, max_seq) + md5
    sock.sendto(packet, (UDP_IP, UDP_PORT))
    endtransmit = time.time()
    print(f'hash cal:{md5.hex()}')
    print(f'Transmission time:{endtransmit-starttransmit}')
    # Schließe den Socket
    
    sock.close()



    


def main():    
    if (len(sys.argv)==7):
        if(int(sys.argv[6])==1):            
            execute_sendSleep(sys.argv[1],sys.argv[2],sys.argv[3],sys.argv[4],sys.argv[5])
        elif(int(sys.argv[6])==2):
            execute_sendACK(sys.argv[1],sys.argv[2],sys.argv[3],sys.argv[4],sys.argv[5])
        elif(int(sys.argv[6])==3):            
            execute_sendSlidingWindow(sys.argv[1],sys.argv[2],sys.argv[3],sys.argv[4],sys.argv[5])
        else:
            print('\nProtocol type not implemented yet')
            sys.exit()    
               
    else : 
        print('\nuse format TransimmionID, Port,IP-Adress,buffersize,Filename with path, protocoltype (1 for sleep 2 for ack 3 for Sliding Window) ')
        sys.exit()




if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print('\naborting transmission')

File: Transmitter/sender.java
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;



public class sender {

    // use java sender.java 1234 5005 localhost 1024 example1MB.txt 

    public static void main(String[] args) {
        if (args.length != 6) {
            System.out.println("Usage: java Sender <transmissionid> <port> <ipaddress> <buffer> <filenameabs> <protocol>");
            return;
        }

        int transmissionId = Integer.parseInt(args[0]);
        int port = Integer.parseInt(args[1]);
        String ipAddress = args[2];
        int buffer = Integer.parseInt(args[3]);
        String filenameAbs = args[4];

        try {
            if(Integer.parseInt(args[5])== 1){
                execute_sendSleep(transmissionId, port, ipAddress, buffer, filenameAbs);
            }else if(Integer.parseInt(args[5])== 2){
                execute_sendACK(transmissionId, port, ipAddress, buffer, filenameAbs);
            }else if(Integer.parseInt(args[5])== 3){
                execute_sendSlidingWindow(transmissionId, port, ipAddress, buffer, filenameAbs); 
            }else System.out.println("Not implemented yet");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void execute_sendSleep(int transmissionId, int port, String ipAddress, int buffer, String filenameAbs)
            throws IOException {
        int bufferSize = buffer;
        String udpIp = ipAddress;
        int udpPort = port;

        String filenameBase = new File(filenameAbs).getName();
                
        DatagramSocket socket = new DatagramSocket();
        InetAddress address = InetAddress.getByName(udpIp);

        int maxSeq = (int) Math.ceil(new File(filenameAbs).length() / (double) bufferSize);
        // letztes paket mit md5
        maxSeq = maxSeq+1;
               
        byte[] header = new byte[10 + filenameBase.length()];
        ByteBuffer.wrap(header, 0, 2).putShort((short) transmissionId);
        ByteBuffer.wrap(header, 2, 4).putInt(0);
        ByteBuffer.wrap(header, 6, 4).putInt(maxSeq);
        System.arraycopy(filenameBase.getBytes(), 0, header, 10, filenameBase.length());

        DatagramPacket packet = new DatagramPacket(header, header.length, address, udpPort);
        socket.send(packet);
              
        
        try (FileInputStream fileInputStream = new FileInputStream(filenameAbs)) {
            byte[] bufferData = new byte[bufferSize];
            int seqNum = 1;
            int packetsSent = 0;

            while (seqNum < maxSeq) {
                int bytesRead = fileInputStream.read(bufferData);
                if (bytesRead == -1) {
                    break;
                }

                ByteBuffer packetData = ByteBuffer.allocate(6 + bytesRead);
                packetData.putShort((short) transmissionId);
                packetData.putInt(seqNum);
                packetData.put(bufferData, 0, bytesRead);

                packet = new DatagramPacket(packetData.array(), packetData.array().length, address, udpPort);
                socket.send(packet);
                packetsSent++;
                try{
                    Thread.sleep((int)(0.0001*bufferSize));
                } catch (InterruptedException e){
                    e.printStackTrace();
                }
                
                seqNum++;
                
                if (maxSeq >= 10 && packetsSent % (maxSeq / 10) == 0) {
                    int percentageSent = Math.round((packetsSent / (float)maxSeq) * 100);
                    System.out.println(percentageSent + "%");
                }
            }

            System.out.println("Transmission complete.");
            
            byte[] md5 = calculateMd5(new File(filenameAbs),bufferSize);
            ByteBuffer md5PacketData = ByteBuffer.allocate(6 + md5.length);
            md5PacketData.putShort((short) transmissionId);
            md5PacketData.putInt(maxSeq);
            md5PacketData.put(md5);

            packet = new DatagramPacket(md5PacketData.array(), md5PacketData.array().length, address, udpPort);
            socket.send(packet);
            System.out.println("Calculated Md5: " + bytesToHex(md5));
            fileInputStream.close();
        }
        socket.close();
    }

    private static void execute_sendACK(int transmissionId, int port, String ipAddress, int buffer, String filenameAbs)
            throws IOException {
        int bufferSize = buffer;
        String udpIp = ipAddress;
        int udpPort = port;

        String filenameBase = new File(filenameAbs).getName();
                
        DatagramSocket socket = new DatagramSocket();
        InetAddress address = InetAddress.getByName(udpIp);

        int maxSeq = (int) Math.ceil(new File(filenameAbs).length() / (double) bufferSize);
        // letztes paket mit md5
        maxSeq = maxSeq+1;
               
        byte[] header = new byte[10 + filenameBase.length()];
        ByteBuffer.wrap(header, 0, 2).putShort((short) transmissionId);
        ByteBuffer.wrap(header, 2, 4).putInt(0);
        ByteBuffer.wrap(header, 6, 4).putInt(maxSeq);
        System.arraycopy(filenameBase.getBytes(), 0, header, 10, filenameBase.length());

        DatagramPacket packet = new DatagramPacket(header, header.length, address, udpPort);
        socket.send(packet);
        
        
        // erhalte erstes ack
        byte[] ackData = new byte[6];
        packet = new DatagramPacket(ackData, ackData.length);
        socket.receive(packet);
        int ackTransId = ByteBuffer.wrap(packet.getData(), 0, 2).getShort();
        int ackSeqNum = ByteBuffer.wrap(packet.getData(), 2, 4).getInt();

        if (ackTransId != transmissionId || ackSeqNum != 0) {
            System.out.println("First packet transfer failed");
            socket.close();
            return;
        }

        try (FileInputStream fileInputStream = new FileInputStream(filenameAbs)) {
            byte[] bufferData = new byte[bufferSize];
            int seqNum = 1;
            int packetsSent = 0;

            while (seqNum < maxSeq) {
                int bytesRead = fileInputStream.read(bufferData);
                if (bytesRead == -1) {
                    break;
                }

                ByteBuffer packetData = ByteBuffer.allocate(6 + bytesRead);
                packetData.putShort((short) transmissionId);
                packetData.putInt(seqNum);
                packetData.put(bufferData, 0, bytesRead);

                packet = new DatagramPacket(packetData.array(), packetData.array().length, address, udpPort);
                socket.send(packet);
                packetsSent++;

                packet = new DatagramPacket(ackData, ackData.length);
                socket.receive(packet);
                ackTransId = ByteBuffer.wrap(packet.getData(), 0, 2).getShort();
                ackSeqNum = ByteBuffer.wrap(packet.getData(), 2, 4).getInt();

                if (ackTransId != transmissionId || ackSeqNum != seqNum) {
                    System.out.println(seqNum + " packet transfer failed");
                    return;
                }

                seqNum++;
                
                if (maxSeq >= 10 && packetsSent % (maxSeq / 10) == 0) {
                    int percentageSent = Math.round((packetsSent / (float)maxSeq) * 100);
                    System.out.println(percentageSent + "%");
                }
            }

            System.out.println("Transmission complete.");
            
            byte[] md5 = calculateMd5(new File(filenameAbs),bufferSize);
            ByteBuffer md5PacketData = ByteBuffer.allocate(6 + md5.length);
            md5PacketData.putShort((short) transmissionId);
            md5PacketData.putInt(maxSeq);
            md5PacketData.put(md5);

            packet = new DatagramPacket(md5PacketData.array(), md5PacketData.array().length, address, udpPort);
            socket.send(packet);
            System.out.println("Calculated Md5: " + bytesToHex(md5));
            fileInputStream.close();
        }
        socket.close();
    }

    private static void execute_sendSlidingWindow(int transmissionId, int port, String ipAddress, int buffer, String filenameAbs)
            throws IOException {
        int bufferSize = buffer;
        String udpIp = ipAddress;
        int udpPort = port;
        final int windowsize=10;

        String filenameBase = new File(filenameAbs).getName();
                
        DatagramSocket socket = new DatagramSocket();
        InetAddress address = InetAddress.getByName(udpIp);

        int maxSeq = (int) Math.ceil(new File(filenameAbs).length() / (double) bufferSize);
        // letztes paket mit md5
        maxSeq = maxSeq+1;
               
        byte[] header = new byte[10 + filenameBase.length()];
        ByteBuffer.wrap(header, 0, 2).putShort((short) transmissionId);
        ByteBuffer.wrap(header, 2, 4).putInt(0);
        ByteBuffer.wrap(header, 6, 4).putInt(maxSeq);
        System.arraycopy(filenameBase.getBytes(), 0, header, 10, filenameBase.length());

        DatagramPacket packet = new DatagramPacket(header, header.length, address, udpPort);
        socket.send(packet);
        
        
        // erhalte erstes ack
        byte[] ackData = new byte[6];
        packet = new DatagramPacket(ackData, ackData.length);
        socket.receive(packet);
        int ackTransId = ByteBuffer.wrap(packet.getData(), 0, 2).getShort();
        int ackSeqNum = ByteBuffer.wrap(packet.getData(), 2, 4).getInt();

        if (ackTransId != transmissionId || ackSeqNum != 0) {
            System.out.println("First packet transfer failed");
            socket.close();
            return;
        }

        // write data into array
        List<byte[]> packet_array= new ArrayList<>();
        int seqNum = 1;
        try (FileInputStream fileInputStream = new FileInputStream(filenameAbs)) {
            byte[] bufferData = new byte[bufferSize];   
            int bytesRead;         
            while ((bytesRead = fileInputStream.read(bufferData)) != -1 && seqNum < maxSeq) {
                
                ByteBuffer packetData = ByteBuffer.allocate(6 + bytesRead);
                packetData.putShort((short) transmissionId);
                packetData.putInt(seqNum);
                packetData.put(bufferData, 0, bytesRead);
                packet_array.add(packetData.array());
                seqNum++;
                //System.out.println("size array"+packet_array.size()+"max seq:"+maxSeq);
            }
            fileInputStream.close();
        }
        
        seqNum=1;
        
        int windowPacketsCounter=0;
        int windowBase=0;
        while (seqNum < maxSeq){
            // sendet window size viele pakete
            while(seqNum<maxSeq && windowPacketsCounter < windowsize){
                packet = new DatagramPacket(packet_array.get(windowBase), packet_array.get(windowBase).length, address, udpPort);
                socket.send(packet);
                seqNum++;
                windowBase++;
                windowPacketsCounter++;
            }
            packet = new DatagramPacket(ackData, ackData.length);
            socket.receive(packet);
            ackTransId = ByteBuffer.wrap(packet.getData(), 0, 2).getShort();
            ackSeqNum = ByteBuffer.wrap(packet.getData(), 2, 4).getInt();  

            if (ackTransId != transmissionId) {
                System.out.println(seqNum + " packet transfer failed");
                socket.close();
                return;
            }


            if(ackSeqNum == seqNum-1){
                System.out.println("alles passst");
            }else if(ackSeqNum < seqNum-1){
                System.out.println("duplicate receiced:"+ackSeqNum+"seq:"+(seqNum-1));
                windowBase=windowBase-(windowBase-ackSeqNum);
                seqNum = windowBase;
            }



            windowPacketsCounter=0;
        }

        System.out.println("Transmission complete.");
            
            byte[] md5 = calculateMd5(new File(filenameAbs),bufferSize);
            ByteBuffer md5PacketData = ByteBuffer.allocate(6 + md5.length);
            md5PacketData.putShort((short) transmissionId);
            md5PacketData.putInt(maxSeq);
            md5PacketData.put(md5);

            packet = new DatagramPacket(md5PacketData.array(), md5PacketData.array().length, address, udpPort);
            socket.send(packet);
            System.out.println("Calculated Md5: " + bytesToHex(md5));
            

        /*
        try (FileInputStream fileInputStream = new FileInputStream(filenameAbs)) {
            byte[] bufferData = new byte[bufferSize];
            int seqNum = 1;
            int packetsSent = 0;

            while (seqNum < maxSeq) {
                int bytesRead = fileInputStream.read(bufferData);
                if (bytesRead == -1) {
                    break;
                }

                ByteBuffer packetData = ByteBuffer.allocate(6 + bytesRead);
                packetData.putShort((short) transmissionId);
                packetData.putInt(seqNum);
                packetData.put(bufferData, 0, bytesRead);

                packet = new DatagramPacket(packetData.array(), packetData.array().length, address, udpPort);
                socket.send(packet);
                packetsSent++;

                packet = new DatagramPacket(ackData, ackData.length);
                socket.receive(packet);
                ackTransId = ByteBuffer.wrap(packet.getData(), 0, 2).getShort();
                ackSeqNum = ByteBuffer.wrap(packet.getData(), 2, 4).getInt();

                if (ackTransId != transmissionId || ackSeqNum != seqNum) {
                    System.out.println(seqNum + " packet transfer failed");
                    return;
                }

                seqNum++;
                
                if (maxSeq >= 10 && packetsSent % (maxSeq / 10) == 0) {
                    int percentageSent = Math.round((packetsSent / (float)maxSeq) * 100);
                    System.out.println(percentageSent + "%");
                }
            }

            System.out.println("Transmission complete.");
            
            byte[] md5 = calculateMd5(new File(filenameAbs),bufferSize);
            ByteBuffer md5PacketData = ByteBuffer.allocate(6 + md5.length);
            md5PacketData.putShort((short) transmissionId);
            md5PacketData.putInt(maxSeq);
            md5PacketData.put(md5);

            packet = new DatagramPacket(md5PacketData.array(), md5PacketData.array().length, address, udpPort);
            socket.send(packet);
            System.out.println("Calculated Md5: " + bytesToHex(md5));
            fileInputStream.close();
        }*/
        socket.close();
    }

    private static byte[] calculateMd5(File file,int bufferSize) {
        try {
            MessageDigest md5 = MessageDigest.getInstance("MD5");
            FileInputStream fileInputStream = new FileInputStream(file);
            byte[] buffer = new byte[bufferSize];
            int bytesRead;
            while ((bytesRead = fileInputStream.read(buffer)) != -1) {
                md5.update(buffer, 0, bytesRead);
            }
            fileInputStream.close();
            return md5.digest();
        } catch (NoSuchAlgorithmException | IOException e) {
            e.printStackTrace();
        }
        return null;
    }
    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}